pkg/runtime.v2/runtime.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2024 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtimev2
+
+import (
+	"errors"
+	"maps"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	kueuelr "sigs.k8s.io/kueue/pkg/util/limitrange"
+
+	kubeflowv2 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v2alpha1"
+)
+
+var (
+	errorDifferentGVK  = errors.New("the GroupVersionKinds are different between old and new objects")
+	errorObjectsAreNil = errors.New("old or new objects are nil")
+)
+
+type Info struct {
+	Obj            client.Object
+	Labels         map[string]string
+	PodLabels      map[string]string
+	Annotations    map[string]string
+	PodAnnotations map[string]string
```

Comment:
Why do we distinguish Labels and PodLabels ?

Reply:
We need to specify some labels only to Pod like scheduler-plugins PodGroup name. And, if we do not distinguish those labels, we need to implement merge JobSet and PodTemplate labels mechanisms...

Reply:
I see, maybe since we don't use PodAnnotations today for Info object, maybe we can remove it for now.

Reply:
I think that we should keep PodLabels and PodAnnotations since users can specify arbitrarily annotations, and we do not restrict the behavior. Actually, the batch/job and JobSet depend on the Pod Annotations.

So, we need to handle those annotations.

- https://github.com/search?q=repo%3Akubernetes-sigs%2Fjobset%20annotations&type=code
- https://kubernetes.io/docs/concepts/workloads/controllers/job/#completion-mode

Reply:
Thos annotations are promoted from the JobSet metadata, isn't ?
Which means they will be used in Annotations of Info object.

Reply:
For example, `alpha.jobset.sigs.k8s.io/node-selector` will be inserted by the JobSet manager, are not promoted from JobSet metadata.

https://github.com/search?q=repo%3Akubernetes-sigs%2Fjobset%20NodeSelectorStrategyKey&type=code

For `batch.kubernetes.io/job-completion-index`, will be inserted by the KubeControllerManager.


In conclusion,  we need to handle annotations properly.

Reply:
I discussed the `PodAnnotations` in the `info` object with @andreyvelich offline.
As a result, we will remove the field for now since it could be useful only for the plugins depending on the Pod annotations since runtimes are immutable and the TrainJob can not be submitted before the runtimes are created.

So, when we support Pod annotations depending on plugins like Yucorn and Volcano or we will make Runtimes as a mutable, we will add the `PodAnnotations` to the `info` object.

Reply:
Done.